File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        mid, ptr = head, head
        while ptr and ptr.next:
            mid = mid.next
            ptr = ptr.next.next

        ptr = mid.next
        rev = mid.next = None
        while ptr:
            next = ptr.next
            ptr.next = rev
            rev = ptr
            ptr = next

        first, second = head, rev
        while second:
            next1, next2 = first.next, second.next
            first.next = second
            second.next = next1
            first, second = next1, next2

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut new_list = None;
        let mut current = head;

        while let Some(mut node) = current.take() {
            current = node.next.take();
            if let Some(new_node) = new_list.take() {
                node.next = Some(new_node);
            }
            new_list = Some(node);
        }

        new_list
    }
}
